ci/prow/make_config.go

```diff
@@ -795,9 +923,65 @@ func strExists(arr []string, str string) bool {
 	return false
 }
 
+type specialBranchLogic struct {
+	branches []string
+	// create new job data based on branches
+	opsNew  func(*baseProwJobTemplateData)
+	restore func(*baseProwJobTemplateData)
+}
+
+// getBase casts data into baseProwJobTemplateData and returns it
+func getBase(data interface{}) *baseProwJobTemplateData {
+	var base *baseProwJobTemplateData
+	switch v := data.(type) {
+	case *presubmitJobTemplateData:
+		base = &data.(*presubmitJobTemplateData).Base
+	case *postsubmitJobTemplateData:
+		base = &data.(*postsubmitJobTemplateData).Base
+	default:
+		log.Fatalf("Unrecognized job template type: '%v'", v)
+	}
+	return base
+}
+
+// recursiveSBL recursively going through specialBranchLogic, and generate job at last
+func recursiveSBL(repoName string, data interface{}, generateOneJob func(data interface{}), sbs []specialBranchLogic, i, posCount int) {
+	// Base case, all special branch logics have been applied
+	if i == len(sbs) {
+		// If there is no branch left, this job shouldn't be generated at all
+		if len(getBase(data).Branches) > 0 || len(getBase(data).SkipBranches) > 0 {
+			generateOneJob(data)
+		}
+		return
+	}
+
+	sb := sbs[i]
+	base := getBase(data)
+
+	origBranches, origSkipBranches := base.Branches, base.SkipBranches
+	// Do legacy branches first
+	base.Branches, base.SkipBranches = consolidateBranches(origBranches, origSkipBranches, sb.branches, []string{})
```

Comment:
This can be an interview question :-)

Reply:
lol. I would fail on this one, it took me a day :(